File: examples/widgets/meter/lv_example_meter_3.py
import utime as time

# Create an image from the png file
try:
    with open('../../assets/img_hand_min.png','rb') as f:
        img_hand_min_data = f.read()
except:
    print("Could not find img_hand_min.png")
    sys.exit()

img_hand_min_dsc = lv.img_dsc_t({
  'data_size': len(img_hand_min_data),
  'data': img_hand_min_data
})

# Create an image from the png file
try:
    with open('../../assets/img_hand_hour.png','rb') as f:
        img_hand_hour_data = f.read()
except:
    print("Could not find img_hand_hour.png")
    sys.exit()

img_hand_hour_dsc = lv.img_dsc_t({
  'data_size': len(img_hand_hour_data),
  'data': img_hand_hour_data
})

def set_value(indic, v):
    meter.set_indicator_value(indic, v)
#
# A clock from a meter
#

def tick_label_event(e):
    draw_part_dsc = e.get_draw_part_dsc();

    # Be sure it's drawing the ticks
    if draw_part_dsc.type != lv.meter.DRAW_PART.TICK: return

    # Be sure it's a major ticks
    if draw_part_dsc.id % 5:  return

    # The order of numbers on the clock is tricky: 12, 1, 2, 3...*/
    txt = ["12", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"]
    # dsc.text is defined char text[16], I must therefore convert the Python string to a byte_array

    idx = int(draw_part_dsc.id / 5)
    draw_part_dsc.text = bytes(txt[idx],"ascii")

meter = lv.meter(lv.scr_act())
meter.set_size(220, 220)
meter.center()

# Create a scale for the minutes
# 60 ticks in a 354 degrees range
meter.set_scale_ticks(60, 1, 10, lv.palette_main(lv.PALETTE.GREY))
meter.set_scale_major_ticks(5, 2, 20, lv.color_black(), 10)         # Every tick is major
meter.set_scale_range(0, 59, 354, 270)

# Add the hands from images
indic_min = meter.add_needle_img(img_hand_min_dsc, 5, 5)
indic_hour = meter.add_needle_img(img_hand_hour_dsc, 5, 5)

#Add an event to set the numbers of hours
meter.add_event(tick_label_event, lv.EVENT.DRAW_PART_BEGIN, None)

# Create an animation to set the value
a1 = lv.anim_t()
a1.init()
a1.set_values(0, 60)
a1.set_repeat_count(lv.ANIM_REPEAT_INFINITE)
a1.set_time(2000)        # 2 sec for 1 turn of the minute hand (1 hour)
a1.set_var(indic_min)
a1.set_custom_exec_cb(lambda a1,val: set_value(indic_min,val))
lv.anim_t.start(a1)

a2 = lv.anim_t()
a2.init()
a2.set_var(indic_hour)
a2.set_time(24000)       # 24 sec for 1 turn of the hour hand
a2.set_values(0, 60)
a2.set_custom_exec_cb(lambda a2,val: set_value(indic_hour,val))
lv.anim_t.start(a2)

File: examples/widgets/meter/lv_example_meter_3.c
#include "../../lv_examples.h"
#if LV_USE_METER && LV_BUILD_EXAMPLES

static lv_obj_t * meter;

static void set_value(void * indic, int32_t v)
{
    lv_meter_set_indicator_end_value(meter, indic, v);
}

static void tick_label_event(lv_event_t * e)
{
    lv_obj_draw_part_dsc_t * draw_part_dsc = lv_event_get_draw_part_dsc(e);

    /*Be sure it's drawing meter related parts*/
    if(draw_part_dsc->class_p != &lv_meter_class) return;

    /*Be sure it's drawing the ticks*/
    if(draw_part_dsc->type != LV_METER_DRAW_PART_TICK) return;

    /*Be sure it's a major ticks*/
    if(draw_part_dsc->id % 5) return;

    /*The order of numbers on the clock is tricky: 12, 1, 2, 3...*/
    if(draw_part_dsc->id == 0) {
        lv_strncpy(draw_part_dsc->text, "12", 4);
    }
    else {
        lv_snprintf(draw_part_dsc->text, 4, "%d", draw_part_dsc->id / 5);
    }
}

/**
 * A clock from a meter
 */
void lv_example_meter_3(void)
{
    meter = lv_meter_create(lv_scr_act());
    lv_obj_set_size(meter, 220, 220);
    lv_obj_center(meter);

    /*Create a scale for the minutes*/
    /*61 ticks in a 360 degrees range (the last and the first line overlaps)*/
    lv_meter_set_scale_ticks(meter, 60, 1, 10, lv_palette_main(LV_PALETTE_GREY));
    lv_meter_set_scale_major_ticks(meter, 5, 2, 20, lv_color_black(), 10);
    lv_meter_set_scale_range(meter, 0, 59, 354, 270);

    LV_IMG_DECLARE(img_hand)

    /*Add a the hands from images*/
    lv_meter_indicator_t * indic_min = lv_meter_add_needle_img(meter, &img_hand, 5, 5);
    lv_meter_indicator_t * indic_hour = lv_meter_add_needle_img(meter, &img_hand, 5, 5);

    lv_obj_add_event_cb(meter, tick_label_event, LV_EVENT_DRAW_PART_BEGIN, NULL);


    /*Create an animation to set the value*/
    lv_anim_t a;
    lv_anim_init(&a);
    lv_anim_set_exec_cb(&a, set_value);
    lv_anim_set_values(&a, 0, 59);
    lv_anim_set_repeat_count(&a, LV_ANIM_REPEAT_INFINITE);
    lv_anim_set_time(&a, 5000);     /*2 sec for 1 turn of the minute hand (1 hour)*/
    lv_anim_set_var(&a, indic_min);
    lv_anim_start(&a);

    lv_anim_set_var(&a, indic_hour);
    lv_anim_set_time(&a, 240000);    /*24 sec for 1 turn of the hour hand*/
    lv_anim_set_values(&a, 0, 59);
    lv_anim_start(&a);
}

#endif
